### PR TITLE
Don't throw away the top bits of our hashed value for strings on 64bi…

### PR DESCRIPTION
**Pull request by aschwaighofer**
#### What's in this pull request?

A fix for String hashValue which was throwing away the top bits from the underlying hash function (i.e NSString.hash) on 64-bit platforms.
#### Resolved bug number: [SR-877](https://bugs.swift.org/browse/SR-877)

Don't throw away the top bits of our hashed value for strings on 64bit platforms

SR-877
rdar://24918173

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

File: validation-test/stdlib/Hashing.swift [hunk 1]
```diff
@@ -128,11 +128,32 @@ HashingTestSuite.test("_squeezeHashValue/UInt") {
...
+#if arch(x86_64) || arch(arm64)
+  // Make sure that we don't accidently throw away bits by storing the result of
+  // NSString.hash into an int in the runtime.
+  var bitSet = false
+  // This is the bit pattern that we xor to NSString's hash value.
+  let hashOffset = UInt(bitPattern: 0x429b_1266_0000_0000)
+  let hash = "efg".hashValue
+  // When we are not equal to the top bit of the xor'ed hashOffset pattern
+  // there where some bits set.
+  let  topHashBits = UInt(bitPattern:hash) & 0xffff_ffff_0000_0000
```
Comment (gribozavr) [inline validation-test/stdlib/Hashing.swift]: Space after the colon, please.

File: validation-test/stdlib/Hashing.swift [hunk 2]
```diff
@@ -128,11 +128,32 @@ HashingTestSuite.test("_squeezeHashValue/UInt") {
...
+  if ((hash > 0) && (topHashBits != hashOffset)) {
```
Comment (gribozavr) [inline validation-test/stdlib/Hashing.swift]: Please drop the extra parens around the condition.

File: validation-test/stdlib/Hashing.swift [hunk 3]
```diff
@@ -128,11 +128,32 @@ HashingTestSuite.test("_squeezeHashValue/UInt") {
...
+    bitSet = true
+  }
+  expectTrue(bitSet)
```
Comment (gribozavr) [inline validation-test/stdlib/Hashing.swift]: You can just do `expectTrue(hash > 0); expectTrue(topHashBits != hashOffset)`

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test OS X platform

File: validation-test/stdlib/Hashing.swift [hunk 4]
```diff
@@ -128,11 +128,30 @@ HashingTestSuite.test("_squeezeHashValue/UInt") {
...
+#if arch(x86_64) || arch(arm64)
+  // Make sure that we don't accidently throw away bits by storing the result of
+  // NSString.hash into an int in the runtime.
+
+  // This is the bit pattern that we xor to NSString's hash value.
+  let hashOffset = UInt(bitPattern: 0x429b_1266_0000_0000)
+  let hash = "efghijkl".hashValue
+  // When we are not equal to the top bit of the xor'ed hashOffset pattern
+  // there where some bits set.
+  let  topHashBits = UInt(bitPattern:hash) & 0xffff_ffff_0000_0000
```
Comment (aschwaighofer) [inline validation-test/stdlib/Hashing.swift]: Missed the space after the colon here. Also whitespace error on the same line. Will fix.

File: validation-test/stdlib/Hashing.swift (hunk 1 above)
Comment (aschwaighofer) [inline validation-test/stdlib/Hashing.swift]: @gribozavr Thanks for taking a look.

**Comment by dduan** [comment]
Nice! This is a tricky little bug.

**Comment by gparker42** [comment]
Which compiler warning do we need to enable to catch this?

File: validation-test/stdlib/Hashing.swift [hunk 5]
```diff
@@ -128,6 +128,24 @@ HashingTestSuite.test("_squeezeHashValue/UInt") {
...
+  let topHashBits = UInt(bitPattern: hash) & 0xffff_ffff_0000_0000
+  expectTrue(hash > 0)
+  expectTrue(topHashBits != hashOffset)
+#endif
+#endif
+}
+
 HashingTestSuite.test("overridePerExecutionHashSeed/overflow") {
   // Test that we don't use checked arithmetic on the seed.
   _HashingDetail.fixedSeedOverride = UInt64.max
```
Comment (Glenn230593) [inline validation-test/stdlib/Hashing.swift]: Leave a line note

**Comment by rjmccall** [comment]
`-Wconversion`

**Pull request merged**